common/src/main/java/feast/common/interceptors/GrpcMessageInterceptor.java

```diff
@@ -0,0 +1,98 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feast.common.interceptors;
+
+import com.google.auto.value.AutoValue;
+import com.google.protobuf.Message;
+import feast.common.logging.AuditLogger;
+import io.grpc.ForwardingServerCall.SimpleForwardingServerCall;
+import io.grpc.ForwardingServerCallListener.SimpleForwardingServerCallListener;
+import io.grpc.Metadata;
+import io.grpc.ServerCall;
+import io.grpc.ServerCall.Listener;
+import io.grpc.ServerCallHandler;
+import io.grpc.ServerInterceptor;
+import org.slf4j.event.Level;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContextHolder;
+
+/**
+ * GrpcMessageInterceptor intercepts a GRPC calls to log handling of GRPC messages to the Audit Log.
+ * Intercepts the incoming and outgoing messages logs them to the audit log, together with method
+ * name and assumed authenticated identity (if authentication is enabled).
+ */
+public class GrpcMessageInterceptor implements ServerInterceptor {
+  /** Collects request and response protobuf messages to allow them to be logged together */
+  @AutoValue
+  public abstract static class MessageCollector {
+    public abstract Message getRequest();
+
+    public abstract Message getResponse();
+
+    @AutoValue.Builder
+    public abstract static class Builder {
+      public abstract Builder setRequest(Message request);
+
+      public abstract Builder setResponse(Message response);
+
+      public abstract MessageCollector build();
+    }
+
+    public static MessageCollector.Builder newBuilder() {
+      return new AutoValue_GrpcMessageInterceptor_MessageCollector.Builder();
+    }
+  }
+
+  @Override
+  public <ReqT, RespT> Listener<ReqT> interceptCall(
+      ServerCall<ReqT, RespT> call, Metadata headers, ServerCallHandler<ReqT, RespT> next) {
+    // Unpack method name from call
+    String fullMethodName = call.getMethodDescriptor().getFullMethodName();
+    String methodName = fullMethodName.substring(fullMethodName.indexOf("/") + 1);
+
+    // Attempt Extract current authenticated identity.
+    Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+    String identity = (authentication == null) ? "" : authentication.getName();
+
+    MessageCollector.Builder messageCollector = MessageCollector.newBuilder();
+    // Register forwarding call to intercept outgoing response and log to audit log
+    call =
+        new SimpleForwardingServerCall<ReqT, RespT>(call) {
+          @Override
+          public void sendMessage(RespT message) {
+            super.sendMessage(message);
+            // 2. Track the response.
+            messageCollector.setResponse((Message) message);
```

Comment:
What happens in the pathological case where no responses are returned and the API is receiving large amount of requests?

Reply:
Adding error handling and make sure that the response would default to a empty Message if not message is returned.